### PR TITLE
document Keycloak authentication

### PR DESCRIPTION
**Pull request by robojones**
Part of Gamify-IT/issues#193

### CODE REVIEW

File: dev-manuals/authentication/authentication.md [hunk 1]
```diff
@@ -1,28 +1,82 @@
...
 
+We are using [Keycloak](https://www.keycloak.org) for user and access management.
+
 When the user logs in, they receive a token.
 The token is stored in a cookie which is sent alongside every request to any backend.
 
-Every backend **must always** validate the token before processing any request.
+## Table of Contents
+
+- [Backend](#backend)
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:10]: ReadTheDocs will hate you.

File: dev-manuals/authentication/authentication.md [hunk 2]
```diff
@@ -1,28 +1,82 @@
...
+@RestController
+class MyRestController {
+    @GetMapping("/")
+    public SomeDAO someRequestHandler(@CookieValue("access_token") String accessToken) {
+        // do something with the accessToken
+    }
+}
+```
+
+Requests without an access token should be rejected with a [401 HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401).
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:32]:
```suggestion
Requests without an access token should be rejected with [401 (Unauthorized)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401).
```

File: dev-manuals/authentication/authentication.md [hunk 3]
```diff
@@ -1,28 +1,82 @@
...
+  - [Access Token](#access-token)
+  - [Validation and User Info](#validation-and-user-info)
+- [Frontend](#frontend)
+- [Keycloak](#keycloak)
+
+## Backend
+
+### Access Token
+
+The access token is sent as a cookie called `access_token` alongside every request.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:20]:
```suggestion
The access token is sent as a cookie called `access_token` alongside every request.
Below you see how to (always) require it for example in `Java` with `Spring`:
```

File: dev-manuals/authentication/authentication.md [hunk 4]
```diff
@@ -1,28 +1,82 @@
 # Authentication and user identification
 
+We are using [Keycloak](https://www.keycloak.org) for user and access management.
+
 When the user logs in, they receive a token.
 The token is stored in a cookie which is sent alongside every request to any backend.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:6]:
```suggestion
When the user logs in, they receive a token. \
The token is stored in a cookie which is sent alongside every request to any backend.
```

File: dev-manuals/authentication/authentication.md [hunk 5]
```diff
@@ -1,28 +1,82 @@
...
+    }
+}
+```
+
+Requests without an access token should be rejected with a [401 HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401).
+
+### Validation and User Info
 
-## Data format
+You can use the access token to get the user's ID, username, roles and other data.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:36]:
```suggestion
You can use the access token to get the user's ID, username, roles and other data. \
```

File: dev-manuals/authentication/authentication.md [hunk 6]
```diff
@@ -1,28 +1,82 @@
...
+The userinfo endpoint of keycloak provides this information.
 
-The cookie has the name `token` and contains a [JSON Web Token](https://jwt.io).
+Locally, the user info endpoint has the URI `http://localhost/keycloak/realms/Gamify-IT/protocol/openid-connect/userinfo`.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:39]:
```suggestion
Locally, the user info endpoint has the URI `http://localhost/keycloak/realms/Gamify-IT/protocol/openid-connect/userinfo`. \
Or, under our reverse proxy under the URI `/keycloak/realms/Gamify-IT/protocol/openid-connect/userinfo`.
```

File: dev-manuals/authentication/authentication.md [hunk 7]
```diff
@@ -1,28 +1,82 @@
...
+You should make this URI configurable in your implementation.
 
-You can validate the token using one of the libraries listed here: https://jwt.io/libraries
+Send a request with the header `Authorization: Bearer <accessToken>` to this endpoint.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:42]:
```suggestion
Send a request with the header `Authorization: Bearer <accessToken>` to this endpoint. \
```

File: dev-manuals/authentication/authentication.md [hunk 8]
```diff
@@ -1,28 +1,82 @@
...
+You then receive a JSON response with the user's information.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:43]:
```suggestion
You then receive a JSON response with the user's information, which looks as follows:
```

File: dev-manuals/authentication/authentication.md [hunk 9]
```diff
@@ -1,28 +1,82 @@
...
+  },
+  "preferred_username": "max",
+  "given_name": "",
+  "family_name": ""
 }
 ```
 
-- `id` uniquely identifies the user. All statistics and other information about a user should be associated with this value.
-- `name` unique username. **Only for debugging.** Even though it is unique, users may be allowed to change it in the future.
+Here a short explanation of the most important fields:
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:63]:
```suggestion
Here is a short explanation of the most important fields:
```

**Review by Gr33ndev — APPROVED** [review]

File: dev-manuals/authentication/authentication.md (hunk 1 above)
Comment (madebyTimo) [reply dev-manuals/authentication/authentication.md:10]: ok I will remove it

**Review by madebyTimo — APPROVED** [review]

File: dev-manuals/authentication/authentication.md [hunk 10]
```diff
@@ -1,28 +1,84 @@
...
-- `name` unique username. **Only for debugging.** Even though it is unique, users may be allowed to change it in the future.
+Here is a short explanation of the most important fields:
+
+- `sub` is the user's ID. It is guaranteed to be unique. It is a string which may not be a UUID.
+- `preferred_username` is the username.
+- `realm_access.roles` contains the user's roles. For students, it includes `student` and for lecturers it includes `lecturer`.
+
+## Frontend
+
+Frontends do not need the user's ID.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:73]:
```suggestion
Frontends do not need the user's ID. \
```

File: dev-manuals/authentication/authentication.md [hunk 11]
```diff
@@ -1,28 +1,84 @@
 # Authentication and user identification
 
-When the user logs in, they receive a token.
+We are using [Keycloak](https://www.keycloak.org) for user and access management.
```
Comment (delvh) [inline dev-manuals/authentication/authentication.md:3]:
```suggestion
We are using [Keycloak](https://www.keycloak.org) for user and access management. \
```

**Review by delvh — APPROVED** [review]

**Pull request merged**